src/lib/traces/tracing-processor.js

```diff
@@ -97,21 +97,119 @@ class TraceProcessor {
+   * @param {number} totalTime Total time (in ms) of interval containing durations.
+   * @param {!Array<number>} percentiles Array of percentiles of interest, in ascending order.
+   * @return {!Array<{percentile: number, time: number}>}
+   * @private
+   */
+  static _riskPercentiles(durations, totalTime, percentiles) {
+    let busyTime = 0;
+    for (let i = 0; i < durations.length; i++) {
+      busyTime += durations[i];
+    }
+
+    // Start with idle time already complete.
+    let completedTime = totalTime - busyTime;
+    let duration = 0;
+    let cdfTime = completedTime;
+    let remainingCount = durations.length + 1;
+    const results = [];
+
+    // Find percentiles of interest, in order.
+    for (let percentile of percentiles) {
+      // Loop over durations, calculating a CDF value for each until it is above
+      // the target percentile.
+      const percentileTime = percentile * totalTime;
+      while (cdfTime < percentileTime && remainingCount > 1) {
+        completedTime += duration;
+        remainingCount--;
+        duration = durations[durations.length - remainingCount];
+
+        // Calculate value of CDF (multiplied by totalTime) for the end of this duration.
+        cdfTime = completedTime + duration * remainingCount;
+      }
+
+      // Negative results are within idle time (0ms wait by definition), so clamp at zero.
+      results.push({
+        percentile,
+        time: Math.max(0, (percentileTime - completedTime) / remainingCount)
+      });
+    }
+
+    return results;
+  }
+
+  /**
+   * Calculates the maximum queueing time (in ms) of high priority tasks for
+   * selected percentiles within a window of the main thread.
+   * @param {!Array<!Object>} trace
+   * @param {number=} startTime Optional start time (in ms) of range of interest. Defaults to trace start.
+   * @param {number=} endTime Optional end time (in ms) of range of interest. Defaults to trace end.
+   * @return {!Array<{percentile: number, time: number}>}
+   */
+  static getRiskToResponsiveness(model, trace, startTime, endTime) {
+    // Range of responsiveness we care about. Default to bounds of model.
+    startTime = startTime === undefined ? model.bounds.min : startTime;
+    endTime = endTime === undefined ? model.bounds.max : endTime;
+    const totalTime = endTime - startTime;
+
+    // Find the main thread.
+    const startEvent = trace.find(event => {
+      return event.name === 'TracingStartedInPage';
+    });
+    const mainThread = TraceProcessor._findMainThreadFromIds(model, startEvent.pid, startEvent.tid);
+
+    // Find durations of all slices in range of interest.
+    // TODO(bckenny): filter for top level slices ourselves?
+    const durations = [];
+    mainThread.sliceGroup.topLevelSlices.forEach(slice => {
+      // Discard slices outside range.
+      if (slice.end <= startTime || slice.start >= endTime) {
+        return;
+      }
+
+      // Clip any at edges of range.
```

Comment:
we can easily not clip, but that can change the question we're asking out from under us.

e.g. you ask for a window of 500ms and theres a 150ms task starting near the end, you're now looking at an up to ~650ms window